### PR TITLE
Use []runtime.Frame instead of exporting Stack (#45)

### PR DESCRIPTION
**Pull request by fgblomqvist**
This is my proposed solution to issue #45. I have tried to preserve as much as possible of the existing code structure to make it a less impactful change, but naturally it will be a breaking one no matter what. I concluded that it made the most sense to use `[]runtime.Frame` instead of e.g. `runtime.Frames` or `[]uintptr`. A runtime.Frame essentially holds the same information as the old `Frame`, with some additions.

With this change in place, it will be easier to extend the package to be able to get stack traces from other error interfaces, as well as to be able to take a stack trace straight (e.g. similar to what https://github.com/stvp/roll allows for).

While the tests are passing, I have yet to test it during some real usage. I'm planning on doing that over the next few days/week, and will report back once I know how it went (assuming this PR is still of interest).

**For now, it depends on #51** simply because that's the kind of environment I've worked in. If that one isn't of interest, it shouldn't be hard to just rebase this onto master (and is something that I'll of course do in that case).

Open to feedback if there is any change/piece of code that isn't up to par/to your liking! :)

### CODE REVIEW

**Comment by fgblomqvist** [comment]
Rebased the PR since #51 got merged.

**Comment by rokob** [comment]
If we are making `BuildStack` private then we need to update some of the comments/docs:

rollbar.go
522:// Callers are required to call BuildStack on their own at the time the cause is wrapped.

doc.go
72:One can implement this interface for custom Error types to be able to build up a chain of stack traces. In order to get stack the correct stacks, callers must call BuildStack on their own at the time that the cause is wrapped. This is the least intrusive mechanism for gathering this information due to the decisions made by the Go runtime to not track this information.

File: transforms.go [hunk 1]
```diff
@@ -191,10 +192,10 @@ func filterIp(ip string, captureIp captureIp) string {
 // method, the causes will be traversed until nil.
 func errorBody(configuration configuration, err error, skip int) (map[string]interface{}, string) {
 	var parent error
-	traceChain := []map[string]interface{}{}
+	var traceChain []map[string]interface{}
```
Comment (rokob) [inline transforms.go]: Why this change?

Comment (fgblomqvist) [reply transforms.go]: Sorry it was just a reflex to change that due to my IDE complaining about declaring an empty slice with a literal. I believe it's considered a good convention to use nil declarations (https://stackoverflow.com/a/29164565/998919), but by all means, I can revert it if this project prefers the other (previous) style.

**Comment by fgblomqvist** [comment]
@rokob I'll do the docs changes (to the best of my knowledge) and push up in a moment, thanks for the suggestion

File: transforms.go (hunk 1 above)
Comment (rokob) [reply transforms.go]: I guess that we always append something to it so this change is okay. The one difference is that for JSON serialization the nil slice becomes `null` and the empty slice becomes `[]` and the backend will treat those differently.

Comment (fgblomqvist) [reply transforms.go]: True. I did pay attention to this in at least one other spot, to make sure we get `[]` in the JSON. Hmm I do feel now that the benefit of leaving that line as-is (e.g. what it originally was) outweighs the benefit of not allocating it (since we do "always" append to it, assuming no errors occur). I'm thinking if the code changes down the line it might become a frustrating error in case a nil slice slips through to the JSON marshaling. I'm gonna go ahead and revert it if that's fine.

File: stack.go [hunk 2]
```diff
@@ -27,21 +27,17 @@ type Frame struct {
...
-type Stack []Frame
+// A stack is a slice of frames
+type stack []frame
 
-// BuildStack uses the Go runtime to construct a slice of frames optionally skipping the number of
-// frames specified by the input skip argument.
-func BuildStack(skip int) Stack {
-	stack := make(Stack, 0)
+// buildStack converts []runtime.Frame into a JSON serializable slice of frames
+// optionally skipping the number of frames specified by the input skip argument.
```
Comment (rokob) [inline stack.go]: You can get rid of the "optionally skipping ..." as we are not taking a skip argument anymore.

**Comment by fgblomqvist** [comment]
@rokob Thanks, done.

**Comment by mmlb** [comment]
Too late now I'd say, but this should have been release as a major version bump.

**Comment by fgblomqvist** [comment]
@mmlb It should. I take part of the blame (I should have suggested that), but yeah, too late now.

**Comment by mmlb** [comment]
:man_shrugging:

**Comment by mmlb** [comment]
Looking forward to updating to this code though :D
